src/util/http/websocket/QueryHub.cpp

```diff
@@ -8,52 +8,55 @@
 
 namespace ad_utility::websocket {
 
-template <bool isSender>
-net::awaitable<std::shared_ptr<
-    QueryHub::ConditionalConst<isSender, QueryToSocketDistributor>>>
-QueryHub::createOrAcquireDistributorInternalUnsafe(QueryId queryId) {
-  while (socketDistributors_->contains(queryId)) {
+net::awaitable<std::shared_ptr<QueryToSocketDistributor>>
+QueryHub::createOrAcquireDistributorInternalUnsafe(QueryId queryId,
+                                                   bool isSender) {
+  if (socketDistributors_->contains(queryId)) {
     auto& reference = socketDistributors_->at(queryId);
     if (auto ptr = reference.pointer_.lock()) {
-      if constexpr (isSender) {
+      if (isSender) {
         // Ensure only single sender reference is acquired for a single session
         AD_CONTRACT_CHECK(!reference.started_);
         reference.started_ = true;
       }
       co_return ptr;
+    } else {
+      socketDistributors_->erase(queryId);
     }
-    // There's the unlikely case where the reference counter reached zero and
-    // the weak pointer can no longer create a shared pointer, but the
-    // destructor is waiting for execution on `globalStrand_`. In this case
-    // re-schedule this coroutine to be executed after destruction. So it is
-    // crucial to use post over dispatch here.
-    co_await net::post(net::bind_executor(globalStrand_, net::use_awaitable));
   }
 
-  auto distributor = std::make_shared<QueryToSocketDistributor>(
-      // We pass a copy of the `shared_pointer socketDistributors_` here,
-      // because in unit tests the callback might be invoked after this
-      // `QueryHub` was destroyed.
-      ioContext_, [&ioContext = ioContext_, globalStrand = globalStrand_,
-                   socketDistributors = socketDistributors_, queryId]() {
-        auto future = net::dispatch(net::bind_executor(
-            globalStrand,
-            std::packaged_task<void()>([&socketDistributors, &queryId]() {
-              bool wasErased = socketDistributors->erase(queryId);
-              AD_CORRECTNESS_CHECK(wasErased);
-            })));
-        // As long as the destructor would have to block anyway, perform work
-        // on the `ioContext_`. This avoids blocking in case the destructor
-        // already runs inside the `ioContext_`.
-        // Note: When called on a strand this may block the current strand.
-        // If the ioContext has been stopped for some reason don't wait
-        // for the result, or this will never terminate.
-        while (future.wait_for(std::chrono::seconds(0)) !=
```

Comment:
There's still a comment in `MessageSender.cpp` about the blocking nature of the destructor, so that should be updated

Reply:
I didn't quite find the comment you meant, but i renamed `distributor` to `distributorAndOwningQueryId` consistently inside the `MessageSender`. That makes this code easier to reason about.